Step2 : (자동차 경주) 문자열 계산기 구현

baek0318 opened the pull request:
### 고민 사항
프로그래밍 요구사항에 메서드가 너무 많은 일을 하지 않도록 분리하려고 노력해라 라고 나와 있는데
제가 생각하기에 메서드가 하나의 일만 하고 있다고 생각하는데 더 나눌 수 있는 부분이 존재할까요?

baek0318 committed:
step2 : kotest 버전 수정

baek0318 committed:
step2 : 사칙연산 계산을 위한 Calculator 코드 작성

baek0318 committed:
step2 : 검증을 위한 테스트 코드 작성

baek0318 committed:
step2 : 오류를 잡아내기 위한 코드 추가
- 비어있는 수식이 입력된 경우
- 숫자가 아닌 문자가 섞여 있는 경우
- 예약된 연산자가 아닌 다른 연산자가 들어온 경우

baek0318 committed:
step2 : 오류를 잡아내는 코드에 대한 검증 테스트 추가

laco-dev commented on build.gradle.kts:
```diff
@@ -13,7 +13,7 @@ repositories {
     testImplementation("org.junit.jupiter", "junit-jupiter", "5.8.2")
     testImplementation("org.assertj", "assertj-core", "3.22.0")
-    testImplementation("io.kotest", "kotest-runner-junit5", "5.2.3")
+    testImplementation("io.kotest", "kotest-runner-junit5", "5.5.0")
```
`kotest 5.5.0`은 클래스 단위의 단위테스트가 skip이 되는 버그가 있습니다
https://kotest.io/docs/changelog.html#551-october-2022

> Tests ignored: 7 of 7 tests

baek0318 replied:
intellij에서 runt test를 gradle을 사용할 경우 해당 문제가 발생하네요
알려주셔서 감사합니다!

laco-dev commented on src/main/kotlin/step2/Calculator.kt:
```diff
@@ -0,0 +1,41 @@
+import java.util.LinkedList
+import java.util.Queue
+
+class Calculator(private val expr: String) {
```
>  사칙연산 계산을 위한 Calculator 코드 작성

라는 커밋으로 기능을 작성 해 주셨지만, 기본적으로는 하나의 기능과 하나의 테스트가 같이 작성되면 좋습니다.
가령 하나의 기능에 문제가 생겨 이 작업을 되돌리려면 다른 기능까지 모두 롤백을 해야 합니다. 또한 이 커밋이 제거되더라도 테스트 코드가 남아있기 때문에 빌드 실패로 이어집니다

작은 단위로 기능을 나누는 연습은 추후 더 하게 되겠지만, 계산기를 구현한다고 가정하면 이런식으로 커밋을 나눌 수도 있겠네요.

> 두개의 숫자를 받아 더하기 연산을 하는 기능을 추가한다

laco-dev replied:
어떤 기능을 구현해야 할지, 작은 기능부터 만들어 가는 연습을 하는 과정이기 때문에 README 파일에다가 기능 요구사항과 프로그래밍 요구사항 등을 정리 해 보시면 어떨까요?

baek0318 replied:
커밋 메세지를 작성할때 롤백까지 고려하지 못했네요 좋은 조언 감사합니다! 다음부터는 좀 더 신경써서 작성하도록 하겠습니다!

README에 정리되면 좀 더 좋을것 같기는 하네요 조언 감사합니다!

laco-dev commented on src/main/kotlin/step2/Calculator.kt:
```diff
@@ -0,0 +1,55 @@
+import java.util.Queue
+
+class Calculator(private val expr: String) {
+    fun calculate(): Int {
```
> 제가 생각하기에 메서드가 하나의 일만 하고 있다고 생각하는데 더 나눌 수 있는 부분이 존재할까요?

Calculator라는 객체 그리고 calculate()라는 함수는 계산을 한다 라는 하나의 일을 한다고 볼 수 있습니다.
하지만 실제로는 (1) 문자열을 파싱하고, (2) 문자열을 분리하고, (3) 연산자와 피연산자를 뽑아내고, (4) 연산을 한다

등의 작은 단위의 일을 여러개 하고 있습니다.
각각 함수로 나누어져 있어서 문제가 없을 수 있지만 추후 작은 기능들의 수정이 생길 때 마다 Calculator의 모든 로직을 다시 확인해야만 합니다.

클린코드 지향을 위한 연습으로 작은 기능들을 클래스로 나누어 보는 연습을 해 보셔도 좋습니다.
1. Calculator
2. Operator
3. ...

힌트 : https://kotlinlang.org/docs/enum-classes.html#implementing-interfaces-in-enum-classes

baek0318 replied:
>각각 함수로 나누어져 있어서 문제가 없을 수 있지만 추후 작은 기능들의 수정이 생길 때 마다 Calculator의 모든 로직을 다시 확인해야만 합니다.

해당 얘기에 공감이 되네요 변경 범위를 최소화 하도록 변경하는게 추후에 유지보수에 훨씬 좋아보이네요 좋은 의견 감사합니다!

laco-dev commented on src/main/kotlin/step2/Calculator.kt:
```diff
@@ -0,0 +1,55 @@
+                .toList()
+        )
+    }
+}
```
> File must end with a newline (\n) (final-newline)
Missing spacing after "while" (keyword-spacing)

`gradlew ktlintCheck` task를 실행 해보세요.
린트 확인을 거친 후에 커밋을 올려주시면 좋습니다 🙏🏻

laco-dev commented on src/main/kotlin/step2/Calculator.kt:
```diff
@@ -0,0 +1,55 @@
+    }
+
+    private fun parse(): Queue<String> {
+        if (expr.isBlank()) {
+            throw IllegalArgumentException("수식이 존재하지 않습니다.")
+        }
```
Calculator 클래스가 생성되면 expr은 불변의 값으로 유지가 됩니다.
하지만 calculate()를 호출할 때 오류가 발생하기 때문에 기능적으로 의미가 없는 클래스가 되어버리는데요.

1. 생성을 할 때 수식이 존재하지 않도록 검사하거나
2. 파라미터를 통해 Calculator 클래스를 재사용할 수 있도록

만들어 보면 어떨까요?

baek0318 replied:
리뷰어님이 얘기한대로 현재 구조에서는 쓸데없는 인스턴스를 만들어서 메모리를 잡아먹는 구조군요..
생성당시에 검사를 하도록 변경을 해야겠네요

laco-dev commented on src/main/kotlin/step2/Calculator.kt:
```diff
@@ -0,0 +1,55 @@
+            if (!isNumber(token) and !isOperator(token)) {
+                throw IllegalArgumentException("수식에 올바른 연산자가 아닙니다.")
+            }
+            if (token == "+") {
+                token = tokens.poll()
+                result += token.toInt()
+            }
+            if (token == "-") {
+                token = tokens.poll()
+                result -= token.toInt()
+            }
+            if (token == "*") {
+                token = tokens.poll()
+                result *= token.toInt()
+            }
+            if (token == "/") {
+                token = tokens.poll()
+                result /= token.toInt()
+            }
```
하나의 토큰을 가져오고 나면, 나머지 조건문은 불필요한 호출 코드가 됩니다.
또한 다른 연산자가 늘어날 때 마다 하드코딩 및 조건문의 분기가 N개씩 늘어나야만 하고 실수로 이어질 수도 있습니다.

아래 코드를 참고 해 보시고, 어떻게 개선할 수 있을지 고민 해보세요
```
val opertor: Operator = Operator.find(token) // 무엇이 반환될까?
val opearnd = ...
result = operator.operate(result, operand)
```

laco-dev commented on src/test/kotlin/step2/CalculatorTest.kt:
```diff
@@ -0,0 +1,53 @@
+
+class CalculatorTest: StringSpec({
+
+    "사칙연산 테스트" {
+        val calculator = Calculator("2 + 3 * 4 / 2")
+        val result = calculator.calculate()
+        result shouldBe 10
+    }
```
테스트가 목적과 의도가 잘 드러나도록 작성 해주셔서 좋네요 👍🏻 
아마도 앞선 의견들을 고려해본다면 테스트가 여러개로 쪼개질 수 있겠습니다

laco-dev commented on src/main/kotlin/step2/Calculator.kt:
```diff
@@ -0,0 +1,55 @@
+package step2
```
이후 미션들은 여러개의 스텝이 하나의 도메인을 담고 있기 때문에 step2 패키지 대신 calculator와 같이 도메인 관점으로 패키지를 나눠주시면 좋습니다!

laco-dev reviewed (changes requested):
좋습니다.
빠르게 구현 해주셨네요. 👍🏻 

본격적으로 자동차 경주 미션 시작하기 전 연습을 겸한 목적으로 몇 가지 의견을 남겨드렸습니다.
잘 고민 해보시고 너무 어렵거나 복잡하다면 시간을 많이 소비하지 말고 적절한 답변을 남겨주시면 됩니다.

baek0318 committed:
step2 : 수식이 없는 에러에 대해서 처리방식 변경
- 인스턴스 생성을 할때 수식이 없으면 에러를 발생 시키도록 변경

baek0318 committed:
step2 : Operator 추가
- 기존에 Calculator에서 처리하던 연산자 판별이나 사칙연산을 담당하는 Operator 컴포넌트 추가

baek0318 committed:
step2 : Expression 추가
- 기존에 Calculator에서 문자열 그대로 들고 있던 부분을 Expression 으로 이동
- 관련 exception 발생도 Expression 내부로 이동
- 수식 파싱 관련 함수도 생성

baek0318 committed:
step2 : README 추가

baek0318 committed:
step2 : Kotest 버전업

baek0318 committed:
step2 : link 적용

baek0318 commented:
@laco-dev 
피드백 주신 것들을 기반으로 리팩토링을 진행하였습니다
리뷰 부탁 드리겠습니다!

### 고민사항
현재 Calculator 내부에 isNumber 이라고 숫자인지 아닌지 판별하는 메서드가 존재하는데 이것만 놓고 보았을때는 별도의 Number를 만들어서 거기서 숫자인지 아닌지를 판별하면 좋겠다라는 생각이 들지만 해당 판별하는 것을 제외하고는 별다른 역할이 없는것 같아서 따로 빼지는 않았습니다 여기에 대한 리뷰어님에 대한 생각이 궁금합니다

laco-dev commented:
> 현재 Calculator 내부에 isNumber 이라고 숫자인지 아닌지 판별하는 메서드가 존재하는데 이것만 놓고 보았을때는 별도의 Number를 만들어서 거기서 숫자인지 아닌지를 판별하면 좋겠다라는 생각이 들지만 해당 판별하는 것을 제외하고는 별다른 역할이 없는것 같아서 따로 빼지는 않았습니다 여기에 대한 리뷰어님에 대한 생각이 궁금합니다

Calculator 입장에서 Number 타입이라는 것에 대해 알아야만 하는가 혹은 알 필요가 없도록 만들것인가에 따라 다를 것 같네요.
가령, 숫자가 아니면 생성 시점에 오류를 발생시키는 객체가 있다면 Calculator 입장에서는 Number인지 검사할 필요가 없게 됩니다.
```kotlin
Number.of("1") // compile error or exception
```

다만 isNumber() 라는 역할만 두고 본다면, 오버엔지니어링처럼 느껴질 수도 있을 것 같네요.
만약 다른 개발자가 isNumber() 라는 동작을 확인할 때마다 매번 Number 클래스를 사용해야 한다는 것을 알 수 있을까..?

아마도 최상위 레벨의 확장 함수같은 유틸 기능으로 만들어지지 않을까도 생각 해 볼만합니다.
다만 계산기 에서는 재사용될 경우가 없기 때문에 특정 클래스의 함수로 작성이 되겠네요.

laco-dev commented on src/main/kotlin/step2/Calculator.kt:
```diff
@@ -4,10 +4,17 @@ import java.util.LinkedList
+    init {
+        if (expr.isBlank()) {
+            throw IllegalArgumentException("수식이 존재하지 않습니다.")
+        }
```
코틀린에서 `IllegalArgumentException` 을 발생시킬 때는 require() 함수를 사용하고는 합니다.
비슷하게는 error() 라는 함수가 있습니다.

laco-dev commented on src/main/kotlin/step2/Operator.kt:
```diff
@@ -0,0 +1,32 @@
+package step2
+
+import java.util.function.IntBinaryOperator
```
예제를 참고해서 오퍼레이터 인터페이스를 구현 해 주셨군요.
코틀린에서는 고차 함수를 이용하여 함수 타입을 활용할 수 있게 됩니다.

차차 학습하면서 활용 해 보시면 좋아요
```kotlin
enum class Operator(
   symbol: String,
   val operate: (Int, Int) -> Int
) {
   PLUS("+") { operand1, operand2 -> operand1 + operand2 },
   ...
}
```

laco-dev commented on src/test/kotlin/step2/OperatorBehaviorSpec.kt:
```diff
@@ -0,0 +1,34 @@
+
+    Given("연산자가 주어진다") {
+        When("연산자가 + 인 경우") {
+            val op = Operator.getOperator("+")
```
`OperatorStringSpec` 에서 "+" 인자로 생성한 타입이 PLUS 라는 것을 검증하고 있어요.
그렇기에 `OperatorBehaviorSpec` 에서는 굳이 getOperator() 를 사용하는 대신 Operator.PLUS를 직접 사용해도 좋습니다.
"+" 라는 수식이 바뀌어버린다면, 다른곳에서도 사용하고 있다는 것을 모르면 테스트 실패로 이어짐

laco-dev commented on src/main/kotlin/step2/Calculator.kt:
```diff
@@ -1,59 +1,24 @@
         while (tokens.isNotEmpty()) {
-            var token = tokens.poll()
-            if (isNumber(token) and !isOperator(token)) {
+            val token = tokens.poll()
+            if (isNumber(token)) {
```
논리적으로 하나의 토큰으로 처리하기 위해 if 문이 반복되었지만, 
계산기에 대한 입력이 정상적으로 들어올 것이다 라고 가정하고 작성한다면

연산자와 피연산자를 한번에 꺼내서 로직을 보다 간단하게 작성할 수도있어요.
애초에 토큰을 꺼내오는데 실패하면 잘못된 수식이 되어버리니까요

```kotlin
val result = tokens.poll()
while (tokens.isNotEmpty()) {
   val operand = tokens.poll()
   val operator = tokens.poll()
}
```

laco-dev reviewed (approved):
피드백 반영 잘 해주셨습니다.

남겨주신 질문 포함해서, 생각 해 볼만한 내용을 몇 가지 남겼습니다.
반영은 꼭 하지 않으셔도 됩니다.

자동차 경주 미션 응원하겠습니다. 💪🏻
